bluemira/equilibria/coils/_grouping.py

```diff
@@ -0,0 +1,1088 @@
+
+
+def symmetrise_coilset(coilset):
+    """
+    Symmetrise a CoilSet by converting any coils that are up-down symmetric about
+    z=0 to SymmetricCircuits.
+
+    Parameters
+    ----------
+    coilset: CoilSet
+        CoilSet to symmetrise
+
+    Returns
+    -------
+    symmetric_coilset: CoilSet
+        New CoilSet with SymmetricCircuits where appropriate
+    """
+    if not check_coilset_symmetric(coilset):
+        bluemira_warn(
+            "Symmetrising a CoilSet which is not purely symmetric about z=0. This can result in undesirable behaviour."
+        )
+    coilset = deepcopy(coilset)
+
+    sym_stack = _get_symmetric_coils(coilset)
+    counts = np.array(sym_stack, dtype=object).T[1]
+
+    new_coils = []
+    for coil, count in zip(coilset._coils, counts):
+        if count == 1:
+            new_coils.append(coil)
+        elif count == 2:
+            if isinstance(coil, SymmetricCircuit):
+                new_coils.append(coil)
+            elif isinstance(coil, Coil):
+                new_coils.append(SymmetricCircuit(coil))
+            else:
+                raise EquilibriaError(f"Unrecognised class {coil.__class__.__name__}")
+        else:
+            raise EquilibriaError("There are super-posed Coils in this CoilSet.")
+
+    return CoilSet(*new_coils)
+
+
+class CoilGroup(CoilGroupFieldsMixin):
+    """
+    Coil Grouping object
+
+    Allow nested coils or groups of coils with access to the methods and properties
+    in the same way to a vanilla Coil
+
+    Parameters
+    ----------
+    coils: Union[Coil, CoilGroup[Coil]]
+
+    """
+
+    def __init__(self, *coils: Union[Coil, CoilGroup[Coil]]):
+        if isinstance(coils[0], Iterable):
+            raise TypeError("Iterable is not a Coil or CoilGroup")
+        self._coils = coils
+        self._pad_discretisation(self.__list_getter("_quad_x"))
+
+    def n_coils(self, ctype: Optional[Union[str, CoilType]] = None) -> int:
+        """
+        Get number of coils
+
+        Parameters
+        ----------
+        ctype: Optional[Union[str, CoilType]]
+            get number of coils of a specific type
+
+        Returns
+        -------
+        int
+            Number of coils
+        """
+        if ctype is None:
+            return len(self.x)
+
+        if not isinstance(ctype, CoilType):
+            ctype = CoilType[ctype]
+
+        return Counter(self.ctype)[ctype]
+
+    def plot(self, ax=None, subcoil=True, label=False, force=None, **kwarg):
+        """
+        Plot a CoilGroup
+        """
+        return CoilGroupPlotter(
+            self, ax=ax, subcoil=subcoil, label=label, force=force, **kwarg
+        )
+
+    def fix_sizes(self):
+        """
+        Fix coils sizes
+        """
+        for fx in self.__list_getter("fix_size"):
+            fx()
+
+    def add_coil(self, *coils: Union[Coil, CoilGroup[Coil]]):
+        """Add coils to the coil group"""
+        self._coils = (*self._coils, *coils)
```

Comment:
I wanted to make it difficult to modify the internal object. If its a tuple most will fail at the first hurdle and use the builtin method. It also means we don't have to do a type conversion when we set the class up as `*coils` makes it already a tuple. I do see your point however. What do you think?

Reply:
It is a private attribute, so users are responsible if they make edits to it. I do get that you're also trying to prevent developers from accidentally editing it, too. It's not too big of a deal as long as `self._coils` is small, if it gets large, then a `list` will be much more efficient.

Reply:
I imagine it will always be a relatively small number of coils and also an uncommon operation (adding or removing coils). Something to keep an eye on I think but I don't think worth changing for now.

Alternatively I could do:

```python
coils_list = list(self._coils)
coils_list.extend(coils)
self._coils = tuple(coils_list)
```